### PR TITLE
#179  Allow user to determine how to handle conflicting output file n…

### PR DESCRIPTION
**Pull request by Hiro-Nakamura**
…ames

### CODE REVIEW

File: public/cli/lib/commands/ffmpeg.ts [hunk 1]
```diff
@@ -8,6 +8,25 @@ import { paths } from '../path-constants';
 
 export async function combineFrames(settings: FfmpegSettings): Promise<void> {
   const executeAudioPath = await combineAudioIfNecessary(settings.audioFiles);
+  
+  // check if output file already exists:
+  var outputName = settings.outputName;
```
Comment (irahopkinson) [inline public/cli/lib/commands/ffmpeg.ts:13]: Don't use `var` anymore (ESLint will complain if you have the extension enabled in VSCode). Use `let` for variables that get reassigned and `const` everywhere else. Here and below.

File: src/App/components/Settings.tsx [hunk 2]
```diff
@@ -162,6 +162,13 @@ export default function Settings(): JSX.Element {
           }}
           onFileSelected={settings.setOutputDirectory}
         />
+        <Checkbox
+          checked={settings.overwriteOutputFiles}
+          onChange={(event): void => {
+            settings.setOverwriteFile(event.currentTarget.checked);
+          }}
+          label="Overwrite existing file(s)"
```
Comment (irahopkinson) [inline src/App/components/Settings.tsx:170]: nit: I wouldn't bother with the parenthesis for plural. Its fine in English to use plural for the general case.

File: public/cli/lib/import/hearThis/hearThisImport.ts [hunk 3]
```diff
@@ -83,10 +83,10 @@ export async function bkImport(project: ConvertProject): Promise<BKProject> {
 // Adapted from https://github.com/caffco/get-audio-duration:
 async function getAudioDurationInMilliseconds(filePath: string): Promise<number> {
   const { output } = spawnSync(paths.ffprobe, ['-show_format', filePath], { stdio: 'pipe' });
-  const matched = output.toString().match(/duration="?(\d*\.\d*)"?/);
+  const matched = (output ?? "").toString().match(/duration="?(\d*\.\d*)"?/);
   if (matched && matched[1]) {
     return parseFloat(matched[1]) * 1000;
   } else {
-    throw new Error('No duration found!');
+    throw new Error(`No duration found! (ffprobe : "${output}")`);
```
Comment (irahopkinson) [inline public/cli/lib/import/hearThis/hearThisImport.ts:90]: Thanks for this improvement to finding errors.

File: public/cli/lib/commands/ffmpeg.ts [hunk 4]
```diff
@@ -8,6 +8,25 @@ import { paths } from '../path-constants';
...
+  if (fs.existsSync(outputName)) {
+    // if they want to keep the file then modify the outputName
+    if (!settings.overwriteOutputFiles) {
+      // get a list of all the files in the directory,
+      var listAllFiles = fs.readdirSync(settings.imagesPath, 'utf8');
+      // get a count of the ones that have a similar start name
+      var nameParts = path.parse(outputName);
+      var matchingFiles = (listAllFiles ??[]).filter((f) => f.indexOf(nameParts.name) == 0);
```
Comment (irahopkinson) [inline public/cli/lib/commands/ffmpeg.ts:21]: In JS you should no longer use `==` for comparison because it coerces and its not always clear what you will get. That is why `===` was introduced. There is one exception which is `x == null` or `x != null` which checks for null or undefined.

Also here you can use the `startsWith` method instead of comparing `indexOf` to zero, which is more readable.

Comment (irahopkinson) [reply public/cli/lib/commands/ffmpeg.ts:21]: I'm not sure if this is doing what we want for the filename but it will do to get started.

**Review by irahopkinson — CHANGES_REQUESTED** [review]
The tests need updating after these changes (that's why the GH build is failing). Run `npm test` and `npm run integration` to run both sets of tests.

**Review by irahopkinson — CHANGES_REQUESTED** [review]
We'll want to rebase this to get the GH build and checks running again. I'll plan to do that either way after PR #190 is merged since there'll probably be things to resolve as a result of that merge.

File: public/cli/lib/commands/convert.ts [hunk 5]
```diff
@@ -139,3 +143,23 @@ function getOutputFilePath(
...
+      // remove the existing file
+      fs.unlinkSync(filePath);
+    } else {
+      // get a list of all the files in the directory,
+      const listAllFiles = fs.readdirSync(path.dirname(outputFilePath), 'utf8');
+      // get a count of the ones that have a similar start name
+      const nameParts = path.parse(filePath);
+      const matchingFiles = (listAllFiles ?? []).filter((f) => f.startsWith(nameParts.name));
+      // since they want to keep the file then modify the filename - new filename = name (length).ext
+      filePath = path.join(nameParts.dir, `${nameParts.name} (${matchingFiles.length})${nameParts.ext}`);
```
Comment (JCFarrow) [inline public/cli/lib/commands/convert.ts:161]: Using the number of similar files probably isn't the best way to rename the file. Creates some cases where the new file isn't saved.
e.g. folder contains: "Mii_Mark_1" and "Mii_Mark_1 (2)"
Maybe make the extension number fill in the gap: "Mii_Mark_1 (1)", or continue after the largest number: "Mii_Mark_1 (3)"

**Review by JCFarrow — CHANGES_REQUESTED** [review]
Looking good. Just a couple of corner cases to fix up.

File: public/cli/lib/commands/convert.ts [hunk 6]
```diff
@@ -139,3 +143,23 @@ function getOutputFilePath(
...
   }
 }
+
+function checkOverwrite(outputFilePath: string, overwriteOutputFiles: boolean): string {
+  let filePath = outputFilePath;
+  // check if output file already exists:
+  if (fs.existsSync(filePath)) {
+    if (overwriteOutputFiles) {
+      // remove the existing file
+      fs.unlinkSync(filePath);
```
[the remark this answers is not in the document]
Comment (irahopkinson) [reply public/cli/lib/commands/convert.ts]: Now that the error reporting is working, I think the right thing to do here is report the error. I don't think we should make the choice for the user to overwrite if they are using the video at the time.

File: public/cli/lib/commands/convert.ts (hunk 5 above)
Comment (irahopkinson) [reply public/cli/lib/commands/convert.ts]: This should be good now. I've added integration tests to prove some of the edge cases.

Comment (JCFarrow) [reply public/cli/lib/commands/convert.ts]: There is still an edge case of a file named "Mii_Mark_1abcdefg (1)" or similar. Which will be sorted to the back of the list and then selected as the highest copy count. Maybe something like this might be better:
```javascript
const copyFilenamePattern = new RegExp("^" + nameParts.name + " \\((\\d+)\\)\\" + nameParts.ext + "$");
```
So that it checks the extension and nothing extra can come before, in between or after.
